salt/metalk8s/orchestrate/bootstrap/init.sls

```diff
@@ -1,23 +1,33 @@
 {# Because of the grain lookup below, the bootstrap minion *must* be available
    before invoking this SLS, otherwise rendering will fail #}
-{%- set bootstrap_control_plane_ip = salt.saltutil.cmd(
-        tgt=pillar.bootstrap_id,
-        fun='grains.get',
-        kwarg={
-            'key': 'metalk8s:control_plane_ip',
-        },
-    )[pillar.bootstrap_id]['ret']
-%}
-
-{%- set bootstrap_workload_plane_ip = salt.saltutil.cmd(
-        tgt=pillar.bootstrap_id,
-        fun='grains.get',
-        kwarg={
-            'key': 'metalk8s:workload_plane_ip',
-        },
-    )[pillar.bootstrap_id]['ret']
-%}
 
+{%- set max_try = 5 %}
+{%- set bootstrap_grains = {} %}
+{%- for _ in range(1, max_try) %}
+  {%- set res = salt.saltutil.cmd(
+          tgt=pillar.bootstrap_id,
+          fun='grains.get',
+          kwarg={
+              'key': 'metalk8s:control_plane_ip',
+          },
+      )
+  %}
+  {%- if pillar.bootstrap_id in res and 'ret' in res[pillar.bootstrap_id] %}
+    {%- do bootstrap_grains.update(
+        {'control_plane_ip': res[pillar.bootstrap_id]['ret']}
+    ) %}
+    {%- break %}
+  {%- endif %}
```

Comment:
Python-style is usually rather "try to get something, and handle failure if it's not there" than "check if something is there, and get it if it is". This avoids duplicating lookups and race conditions (although there is no risk of such condition in your case).

```suggestion
  {%- set control_plane_ip = res.get(pillar.bootstrap_id, {}).get('ret') %}
  {%- if control_plane_ip is not none %}
    {%- do bootstrap_grains.update({'control_plane_ip': control_plane_ip}) %}
    {%- break %}
  {%- endif %}
```